docs_src/design/adr/0027-URIs for Files.md

```diff
@@ -0,0 +1,122 @@
+
+### Authentication 
+
+#### username-password in URI
+
+While not recommended, users will be able to specify **username-password** (`<username>:<password>@`) in the URI in plain text. While this is ok network wise when using HTTPS, it isn't good practice to have these credentials specified in configuration or other service files where the URI is specified.
+
+!!! example - "Example plain text `username-password` in URI located in configuration"
+    ```toml
+    [UoM]
+    UoMFile = "https://myuser:mypassword@example.com/uom.yaml"
+    ```
+
+#### Secure Credentials
+
+In order to provide a secure way for users to specify credentials, the `edgexSecretName` query parameter can be specified on the URI. This parameter specifies a Secret Name from the service's Secret Store where the credentials reside and will be processed by the new helper function.
+
+!!! example - "Example URI with `edgexSecretName` query parameter"
+    ```toml
+    [UoM]
+    UoMFile = "https://example.com/uom.yaml?edgexSecretName=mySecretName"
+    ```    
+
+The type of authentication as well as the credentials will be contained in the secret data specified by the Secret Name. Two types of authentication will be supported, which are `usernamepassword`and `httpheader`. The `httpheader`type  will accommodate various forms of an `Authorization` header as well as typical API key usages
+
+- When `usernamepassword` is specified as the type in the secret data, the `username:password@` will be inserted into the URI using the **username** and **password** found in the secret data.
+
+    !!! example - "Example secret data - `usernamepassword`"
+        ```
+        type=usernamepassword
+        username=myuser
+        password=mypassword
+        ```
+        Resulting URI with `basic-auth`"
+        ```
+        https://myuser:mypassword@example.com/uom.yaml
+        ```
+        
+- When `httpheader` is specified as the type in the secret data, the header name and contents from the secret data  will be placed in the HTTP header. 
+
+    !!! example - "Example secret data - `Authorization` using  `httpheader`"
+        ```
+        type=httpheader
+        headername=Authorization
+        headercontents=Basic bXl1c2VyOm15cGFzc3dvcmQ=
+        ```
+        For a request header set as:
+        ```
+        GET https://example.com/uom.yaml HTTP/1.1
+        Authorization: Basic bXl1c2VyOm15cGFzc3dvcmQ=
+        ```

```

Comment:
Given the changes to allow setting custom headers, these two mechanisms have become redundant. The credentials passed in the URI are just a common mechanism for when setting the header isn't possible for basic auth. The user can base64-encode the `user:pass` and set it as the `Authorization: Basic <base64-encoded>`.

As @bnevis-i suggested, one type is sufficient to cover all header-based authentication.

See what curl does with credentials passed in URL:
```
$ curl -v http://user:pass@example.com
*   Trying 93.184.216.34:80...
*   Trying 2606:2800:220:1:248:1893:25c8:1946:80...
* Connected to example.com (93.184.216.34) port 80 (#0)
* Server auth using Basic with user 'user'
> GET / HTTP/1.1
> Host: example.com
> Authorization: Basic dXNlcjpwYXNz  <------- The creds are sent to the server in header
> User-Agent: curl/7.81.0
> Accept: */*
> 
* Mark bundle as not supporting multiuse
< HTTP/1.1 200 OK
...
```

Reply:
With that said, I am fine with keeping raw username and password for basic auth for the sake of usability. But the type should be called specifically to indicate what it is: `basic-auth`. Moreover, we shouldn't put a basic auth example with custom header because we already have basic auth supported by accepting raw user/pass.

Reply:
@farshidtz   I wanted to point out that a potential benefit of username password vs a raw header is that if the remote server requires digest auth, knowing the username and password would allow use of digest auth, whereas a raw header wouldn't.

Reply:
Makes sense. But the example should NOT be:
```
        Resulting URI with `basic-auth`"
        ```
        https://myuser:mypassword@example.com/uom.yaml
        ```
```
Because in case of Basic Auth, the we should pass the given credentials in header as it gives lower chance of ending up in logs.

Also, the example in the case of `httpheader` should not show the basic auth, but rather a Bearer token example.

Reply:
Based in this discussion, I've added suggestion below.

I kept the basic auth example for custom header, in case direct basic auth is needed without asking if server needs it.

Reply:
> Makes sense. But the example should NOT be:

@farshidtz I agree.  Don't change the URI; just do the auth.  

> Also, the example in the case of httpheader should not show the basic auth, but rather a Bearer token example.

It should be an API key example.  Bearer tokens are supposed to have a short lifetime.  It's unlikely anyone interacting with a real system would have long-term success with a static bearer token.  Though technically it is possible.

Reply:
Many servers use OAuth2 and in turn, need to pass a Bearer token. It is true that Bearer tokens should be short lived, but that doesn't prevent an application from refreshing that short-lived token in the EdgeX Secret Store to allow continuous access to the file server. Anyway, this is just an example and it would be technically possible regardless of the example.

Reply:
@farshidtz , @bnevis-i and I discussed this and agreed that `httpheader` covers all needs including basic-auth. We will drop `usernamepassword` and only have `httpheader`. Examples will be updated to show `api-key`, `basic-auth`, and `Bearer`. At this time digest will not be supported. I can be added at a later time base on feedback of its need.

Reply:
Updates have been pushed